pkg/controllers/machine/link/controller.go

```diff
@@ -0,0 +1,129 @@
+	cache         *cache.Cache // exists due to eventual consistency on the controller-runtime cache
+}
+
+func NewController(kubeClient client.Client, cloudProvider *cloudprovider.CloudProvider) controller.Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+		cache:         cache.New(time.Minute, time.Second*10),
+	}
+}
+
+func (c *Controller) Name() string {
+	return "machine.link"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	machineList := &v1alpha5.MachineList{}
+	if err := c.kubeClient.List(ctx, machineList); err != nil {
+		return reconcile.Result{}, err
+	}
+	retrieved, err := c.cloudProvider.List(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing cloudprovider machines, %w", err)
+	}
+	// Filter out any machines that shouldn't be linked
+	retrieved = lo.Filter(retrieved, func(m *v1alpha5.Machine, _ int) bool {
+		return c.shouldLink(m, machineList.Items)
+	})
+	errs := make([]error, len(retrieved))
+	workqueue.ParallelizeUntil(ctx, 20, len(retrieved), func(i int) {
+		errs[i] = c.link(ctx, retrieved[i])
+	})
+	// Effectively, don't requeue this again once it succeeds
+	return reconcile.Result{RequeueAfter: math.MaxInt64}, multierr.Combine(errs...)
+}
+
+func (c *Controller) link(ctx context.Context, retrieved *v1alpha5.Machine) error {
+	provisionerName, ok := retrieved.Labels[v1alpha5.ProvisionerNameLabelKey]
+	if !ok {
+		return corecloudprovider.IgnoreMachineNotFoundError(c.cloudProvider.Delete(ctx, retrieved))
+	}
+	provisioner := &v1alpha5.Provisioner{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: provisionerName}, provisioner); err != nil {
+		if errors.IsNotFound(err) {
+			return corecloudprovider.IgnoreMachineNotFoundError(c.cloudProvider.Delete(ctx, retrieved))
+		}
+		return err
+	}
+	machine := machineutil.New(&v1.Node{}, provisioner)
+	machine.GenerateName = fmt.Sprintf("%s-", provisioner.Name)
+	// This annotation communicates to the machine controller that this is a machine linking scenario, not
+	// a case where we want to provision a new machine
+	machine.Annotations = lo.Assign(machine.Annotations, map[string]string{
+		v1alpha5.MachineLinkedAnnotationKey: retrieved.Status.ProviderID,
+	})
+	if err := c.kubeClient.Create(ctx, machine); err != nil {
+		return err
+	}
+	c.cache.SetDefault(retrieved.Status.ProviderID, nil)
+	return nil
+}
+
+// shouldLink checks if the cloudprovider machine is owned by a provisioner but if it wasn't created
+// by a machine. If both of these are true, then we should generate a Machine
+func (c *Controller) shouldLink(retrieved *v1alpha5.Machine, existingMachines []v1alpha5.Machine) bool {
+	if _, ok := retrieved.Labels[v1alpha5.ManagedByLabelKey]; ok {
+		return false
+	}
+	// Machine was already linked but controller-runtime cache didn't update
+	if _, ok := c.cache.Get(retrieved.Status.ProviderID); ok {
```

Comment:
Is there harm in just linking again rather than yet-another-cache?

Reply:
I think so since we would get two Machines that have the same provider id. There may be room to clean this up in a garbage collection flow but I'd like to avoid it for now if we can. Plus, the cache shouldn't be that expensive from a memory perspective and we should be ripping this code out at some point when the migration is done